src/NuGet.Core/NuGet.PackageManagement/NuGetPackageManager.cs

```diff
@@ -2488,41 +2579,71 @@ var projectUniqueNamesForBuildIntToUpdate
                 return null;
             }
 
-            var stopWatch = Stopwatch.StartNew();
-
-            // Find all sources used in the project actions
-            var sources = new HashSet<SourceRepository>(
-                nuGetProjectActions.Where(action => action.SourceRepository != null)
-                    .Select(action => action.SourceRepository),
-                    new SourceRepositoryComparer());
+            var resolvedAction = await PreviewBuildIntegratedProjectsActionsAsync(
+                new List<BuildIntegratedNuGetProject>() { buildIntegratedProject },
+                new Dictionary<string, IEnumerable<NuGetProjectAction>>(PathUtility.GetStringComparerBasedOnOS())
+                {
+                    { buildIntegratedProject.MSBuildProjectPath, nuGetProjectActions}
+                },
+                packageIdentity: null, // since we have nuGetProjectActions no need packageIdentity
+                primarySources: null, // since we have nuGetProjectActions no need primarySources
+                nuGetProjectContext,
+                token
+                );
 
-            // Add all enabled sources for the existing packages
-            var enabledSources = SourceRepositoryProvider.GetRepositories();
+            return resolvedAction.FirstOrDefault(r => r.Project == buildIntegratedProject)?.Action as BuildIntegratedProjectAction;
+        }
 
-            sources.UnionWith(enabledSources);
 
-            // Read the current lock file if it exists
-            LockFile originalLockFile = null;
-            var lockFileFormat = new LockFileFormat();
+        /// <summary>
+        /// Run project actions for build integrated many projects.
+        /// </summary>
+        internal async Task<IEnumerable<ResolvedAction>> PreviewBuildIntegratedProjectsActionsAsync(
```

Comment:
Why not private method?

Reply:
Sure, I have it private. We need to try from lowest modifier level to higher.